File: javascript/command_injection/systemUtils.js
const { exec, execSync } = require('child_process');

// VULN 1: exec with string concatenation - ping
function checkHostAvailability(hostname) {
    exec('ping -c 4 ' + hostname, (err, stdout) => {
        return !err;
    });
}

// VULN 2: execSync with template literal - traceroute
function traceNetworkRoute(destination) {
    return execSync(`traceroute ${destination}`).toString();
}

// VULN 3: execSync with string concatenation - DNS lookup
function resolveHostname(hostname) {
    return execSync('nslookup ' + hostname).toString();
}

module.exports = { checkHostAvailability, traceNetworkRoute, resolveHostname };

File: go/command_injection/system_utils.go
package commandinjection

import (
	"fmt"
	"os/exec"
)

// VULN 1: exec.Command sh -c with string concatenation - ping
func CheckHostAvailability(hostname string) (bool, error) {
	cmd := exec.Command("sh", "-c", "ping -c 4 "+hostname)
	err := cmd.Run()
	return err == nil, err
}

// VULN 2: exec.Command sh -c with fmt.Sprintf - traceroute
func TraceNetworkRoute(destination string) ([]byte, error) {
	cmdStr := fmt.Sprintf("traceroute %s", destination)
	cmd := exec.Command("sh", "-c", cmdStr)
	return cmd.Output()
}

// VULN 3: exec.Command sh -c with string concatenation - nslookup
func ResolveHostname(hostname string) ([]byte, error) {
	cmd := exec.Command("sh", "-c", "nslookup "+hostname)
	return cmd.Output()
}
